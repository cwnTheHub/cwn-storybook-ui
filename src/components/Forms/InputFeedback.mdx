import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Story,
} from "@storybook/blocks";

import * as InputFeedbackStories from "./InputFeedback.stories";

<Meta of={InputFeedbackStories} />
<Title />

`import { InputFeedback } from '@nds_core/nds'`

### PROPS & METHODS

<Controls />

### Standalone usage
While its primary use is to facilitate feedback states for other form components such as Input, you may use it standalone.

<Canvas of={InputFeedbackStories.Standalone} />

### Advanced use with form components

Passing a function that returns an InputFeedback to the helper prop of form components gives you more control over the contents of the InputFeedback component by giving you access to the feedback state of the form field and the current value.

As an example, enter a value into the field below, then click away to lose focus. If you enter less than the 16 character minimum the helper will show as an error. Enter 16 or more characters to receive the success feedback.