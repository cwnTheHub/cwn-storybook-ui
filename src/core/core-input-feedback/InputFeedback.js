import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { borders, messaging } from "../../shared-styles";
import Box from "../core-box/Box";
import { safeRest } from "../../util-helpers";

const StyledFeedback = styled(({ feedback, ...rest }) => <Box {...rest} />)(
  ({ feedback }) => ({
    ...borders.rounded,
    ...(feedback === "success" && messaging.success),
    ...(feedback === "error" && messaging.error),
    ...(feedback === undefined && messaging.standard),
  })
);

const InputFeedback = ({ feedback, children, ...rest }) => (
  <StyledFeedback
    {...safeRest(rest)}
    inset={3}
    role={feedback === "error" ? "alert" : null}
    feedback={feedback}
  >
    {children}
  </StyledFeedback>
);

InputFeedback.propTypes = {
  feedback: PropTypes.oneOf(["success", "error"]),
  children: PropTypes.node.isRequired,
};

InputFeedback.defaultProps = {
  feedback: undefined,
};

export default InputFeedback;
