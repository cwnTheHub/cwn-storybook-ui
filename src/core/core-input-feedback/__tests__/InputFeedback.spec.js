import React from "react";
import { shallow, mount, render } from "enzyme";

import InputFeedback from "../InputFeedback";

describe("InputFeedback", () => {
  const defaultChildren = "Some feedback text.";
  const doShallow = (props = {}, children = defaultChildren) =>
    shallow(<InputFeedback {...props}>{children}</InputFeedback>);

  const doMount = (props = {}, children = defaultChildren) => {
    const inputFeedback = mount(
      <InputFeedback {...props}>{children}</InputFeedback>
    );

    return inputFeedback.find("div");
  };

  it("renders", () => {
    const inputFeedback = render(
      <InputFeedback>Some inputFeedback text.</InputFeedback>
    );

    expect(inputFeedback).toMatchSnapshot();
  });

  it("can have a feedback state", () => {
    let inputFeedback = doMount();
    expect(inputFeedback).toMatchSnapshot();

    inputFeedback = doMount({ feedback: "success" });
    expect(inputFeedback).toMatchSnapshot();
  });

  it("passes additional attributes to the element", () => {
    const inputFeedback = doShallow({ id: "the-inputFeedback" });

    expect(inputFeedback.props().id).toEqual("the-inputFeedback");
  });

  it("sets role to alert on error", () => {
    const inputFeedback = doShallow({ feedback: "error" });

    expect(inputFeedback.props().role).toEqual("alert");
  });

  it("does not allow custom CSS", () => {
    const inputFeedback = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(inputFeedback.props()["className"]).not.toEqual("my-custom-class");
    expect(inputFeedback.props()["style"]).toBe(undefined);
  });
});
