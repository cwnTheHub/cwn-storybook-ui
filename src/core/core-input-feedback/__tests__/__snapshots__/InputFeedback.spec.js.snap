// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFeedback can have a feedback state 1`] = `
<div
  className="sc-aXZVg hNbjKD sc-gEvEer jUueUw sc-gEvEer jUueUw"
  role={null}
>
  Some feedback text.
</div>
`;

exports[`InputFeedback can have a feedback state 2`] = `
<div
  className="sc-aXZVg hNbjKD sc-gEvEer hTQEAH sc-gEvEer hTQEAH"
  role={null}
>
  Some feedback text.
</div>
`;

exports[`InputFeedback renders 1`] = `
<div
  class="sc-aXZVg hNbjKD sc-gEvEer jUueUw sc-gEvEer jUueUw"
>
  Some inputFeedback text.
</div>
`;
